src/content/contribute/writing-a-plugin.md

```diff
@@ -182,40 +183,94 @@ module.exports = HelloAsyncPlugin;
 
 Once we can latch onto the webpack compiler and each individual compilations, the possibilities become endless for what we can do with the engine itself. We can reformat existing files, create derivative files, or fabricate entirely new assets.
 
-Let's write a simple example plugin that generates a new build file called `filelist.md`; the contents of which will list all of the asset files in our build. This plugin might look something like this:
+Let's write a simple example plugin that generates a new build file called `assets.md`, the contents of which will list all of the asset files in our build. This plugin might look something like this:
 
 ```javascript
 class FileListPlugin {
+
+  static defaultOptions = {
+    outputFile: 'assets.md',
+  };
+
+  // Any options should be passed in the constructor of your plugin,
+  // (this is a public API of your plugin)
+  constructor(options = {}) {
+    // Applying user-specified options over the default options
+    // and making merged options further available to the plugin methods.
+    // You should probably validate all the options here as well
+    this.options = { ...FileListPlugin.defaultOptions, options };
+  }
+
   apply(compiler) {
-    // emit is asynchronous hook, tapping into it using tapAsync, you can use tapPromise/tap(synchronous) as well
-    compiler.hooks.emit.tapAsync('FileListPlugin', (compilation, callback) => {
-      // Create a header string for the generated file:
-      var filelist = 'In this build:\n\n';
-
-      // Loop through all compiled assets,
-      // adding a new line item for each filename.
-      for (var filename in compilation.assets) {
-        filelist += '- ' + filename + '\n';
-      }
 
-      // Insert this list into the webpack build as a new file asset:
-      compilation.assets['filelist.md'] = {
-        source: function () {
-          return filelist;
-        },
-        size: function () {
-          return filelist.length;
-        },
-      };
-
-      callback();
+    // webpack module instance can be accessed from the compiler object,
+    // this ensures that correct version of the module is used
+    const { webpack } = compiler;
+
+    // RawSource is one of the "sources" classes that should be used
+    // to represent asset sources in compilation
+    const { RawSource } = webpack.sources;
+
+    // Tapping to the "emit" hook
+    compiler.hooks.emit.tap(FileListPlugin.name, (compilation) => {
```

Comment:
should use `processAssets` hook with right stage, look at `copy-webpack-plugin`

Reply:
Yeah, I've seen the warning in the console today. I will look into it, thanks.